In [1]:
import warnings
warnings.simplefilter('ignore')

# Marvin Queries
This tutorial goes through a few basics of how to perform queries on the MaNGA dataset using the Marvin Query tool. Please see the [Marvin Query](https://sdss-marvin.readthedocs.io/en/latest/query/query.html) page for more details on how to use Queries.  This tutorial covers the basics of:


 * querying on metadata information from the NSA catalog
 * how to combine multiple filter and return additional parameters
 * how to perform radial cone searches with Marvin
 * querying on information from the MaNGA DAPall summary file
 * querying using quality and target flags 

First let's import some basics

In [4]:
# we should be using DR15 MaNGA data
from marvin import config
config.release

# import the Query tool
from marvin.tools.query import Query

INFO: No release version set. Setting default to DR15


/Users/Brian/anaconda2/envs/marvindev/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/Users/Brian/anaconda2/envs/marvindev/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/Users/Brian/anaconda2/envs/marvindev/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)


## Query Basics  
### Querying on Metadata 
Let's go through some Query basics of how to do a query on metadata.  The two main keyword arguments to Query are **search_filter** and **return_params**.  **search_filter** is a string representing the SQL ``where`` condition you'd like to filter on.  This tutorial assumes a basic familiarity with the SQL boolean syntax needed to construct Marvin Queries.  Please see the [tutorial on SQL Boolean syntax](https://sdss-marvin.readthedocs.io/en/latest/tutorials/boolean-search-tutorial.html) to learn more. **return_params** is a list of parameters you want to return in the query in addition to those used in the SQL filter condition. 

Let's search for all galaxies with a redshift less than 0.1. To specify our search parameter, redshift, we must know the database table and name of the parameter. In this case, MaNGA uses the NASA-Sloan Atlas (NSA) for redshift information.  In the NSA catalog, the redshift is the **z** parameter of the **nsa** table, so our search parameter will be ``nsa.z``.  Generically, all search parameters will take the form `table.parameter`.

In [5]:
# filter for galaxies with a redshift < 0.1
my_filter = 'nsa.z < 0.1'

In [7]:
# construct the query
q = Query(search_filter=my_filter)
q

Marvin Query(filter='nsa.z < 0.1', mode='remote', data_origin='api')

The `Query` tool works with a local or remote manga database.  Without a database, the ``Query`` tool submits your inputs to the remote server using the API, rather than doing anything locally.  We run the query with the ``run`` method.  Your inputs are sent to the server where the query is built dynamically and run.  The results of the query are returned via the API, parsed and converted into a Marvin ``Results`` object.  

In [8]:
# run the query
r = q.run()

# print some results information
print(r)
print('number of results:', r.totalcount)

Marvin Results(query=nsa.z < 0.1, totalcount=4275, count=100, mode=remote)
number of results: 4275


After constructing queries, we can run them with **q.run()**.  This returns a **Marvin Results** object. Let's take a look.  This query returned 4275 objects.  For queries with large results, the results are automatically paginated in sets of 100 objects.  Default parameters returned in queries always include the **mangaid** and **plateifu**.  Marvin Queries will also return any parameters used in the definition of your filter condition. Since we filtered on redshift, the redshift is automatically included.

In [12]:
# look at the current page of results (subset of 10)
print('number in current set:', len(r.results))
print(r.results[0:10])

number in current set: 100
<ResultSet(set=1.0/428, index=0:10, count_in_set=10, total=4275)>
[ResultRow(mangaid='1-109056', plateifu='8077-6103', z=0.0473019),
 ResultRow(mangaid='1-109073', plateifu='8078-12704', z=0.0435882),
 ResultRow(mangaid='1-109081', plateifu='8077-12705', z=0.0438592),
 ResultRow(mangaid='1-109112', plateifu='8078-1901', z=0.0249702),
 ResultRow(mangaid='1-109152', plateifu='8154-12702', z=0.0281736),
 ResultRow(mangaid='1-109167', plateifu='8078-1902', z=0.0246344),
 ResultRow(mangaid='1-109234', plateifu='9193-1901', z=0.0409309),
 ResultRow(mangaid='1-109244', plateifu='8080-12702', z=0.0267911),
 ResultRow(mangaid='1-109250', plateifu='8080-1902', z=0.0383731),
 ResultRow(mangaid='1-109270', plateifu='9193-12703', z=0.0252266)]


### Finding Available Parameters
We can use the ``Query`` datamodel to look up all the available parameters one can use in the ``search_filter`` or in the ``return_params`` keyword arguments to ``Query``.  The ``Query`` datamodel contains many **parameters**, which we bundle into **groups** for easier navigation.  We currently offer four groups of parameters, with remaining parametrs captured in an **Other** group.  See the [Query Datamodel](https://sdss-marvin.readthedocs.io/en/latest/datamodel/query_dm.html) in the docs for more detailed info.

In [10]:
# look up the available query datamodel groups
q.datamodel.groups

[<ParameterGroup name=Metadata, n_parameters=11>,
 <ParameterGroup name=NSA Catalog, n_parameters=158>,
 <ParameterGroup name=ObsInfo, n_parameters=64>,
 <ParameterGroup name=DAPall Summary, n_parameters=620>,
 <ParameterGroup name=Other, n_parameters=134>]

Each group contains a list of ``QueryParameters``.  A ``QueryParameter`` contains the designation and syntax used for specifying parameters in your query.  The ``full`` attribute indicates the **table.parameter** name needed as input into the ``search_filter`` or ``return_params`` keyword arguments to the Marvin ``Query``.  The redshift is a parameter from the NSA catalog.  There are 158 parameters in the **NSA Catalog** group. 

In [9]:
# select and print the NSA parameter group
nsa = q.datamodel.groups['nsa']
nsa.parameters

[<QueryParameter full=nsa.iauname, name=iauname, short=iauname, remote=iauname, display=IAU Name>,
 <QueryParameter full=nsa.ra, name=ra, short=ra, remote=ra, display=RA>,
 <QueryParameter full=nsa.dec, name=dec, short=dec, remote=dec, display=Dec>,
 <QueryParameter full=nsa.z, name=z, short=z, remote=z, display=Redshift>,
 <QueryParameter full=nsa.elpetro_ba, name=elpetro_ba, short=axisratio, remote=elpetro_ba, display=Elpetro axis ratio>,
 <QueryParameter full=nsa.elpetro_mag_g_r, name=elpetro_mag_g_r, short=g_r, remote=elpetro_mag_g_r, display=g-r>,
 <QueryParameter full=nsa.elpetro_absmag_g_r, name=elpetro_absmag_g_r, short=absmag_g_r, remote=elpetro_absmag_g_r, display=Absmag g-r>,
 <QueryParameter full=nsa.elpetro_logmass, name=elpetro_logmass, short=logmass, remote=elpetro_logmass, display=Elpetro Stellar Mass>,
 <QueryParameter full=nsa.elpetro_th50_r, name=elpetro_th50_r, short=th50_r, remote=elpetro_th50_r, display=r-band half-light radius>,
 <QueryParameter full=nsa.sersic_l

### Multiple Search Criteria and Returning Additional Parameters
We can easily combine query filter conditions by constructing a boolean string using AND.  Let's search for galaxies with a redshift < 0.1 and log M$_\star$ < 10.  The NSA catalog contains the Sersic profile determination for stellar mass, which is the **sersic_mass** or **sersic_logmass** parameter of the **`nsa`** table, so our search parameter will be **nsa.sersic_logmass**.  

Let's also return the object RA and Dec as well using the **return_params** keyword.  This accepts a list of string parameters.  Object RA and Dec are included in the **cube** table so the parameter names are `cube.ra` and `cube.dec`.

In [13]:
my_filter = 'nsa.z < 0.1 and nsa.sersic_logmass < 10'
q = Query(search_filter=my_filter, return_params=['cube.ra', 'cube.dec'])
r = q.run()
print(r)
print('Number of objects:', r.totalcount)

Marvin Results(query=nsa.z < 0.1 and nsa.sersic_logmass < 10, totalcount=1932, count=100, mode=remote)
Number of objects: 1932


This query return 1932 objects and now includes the RA, Dec, redshift and log Sersic stellar mass parameters. 

In [14]:
# print the first 10 rows
r.results[0:10]

<ResultSet(set=1.0/194, index=0:10, count_in_set=10, total=1932)>
[ResultRow(mangaid='1-109073', plateifu='8078-12704', ra=42.1567757009, dec=-0.554292603774, sersic_logmass=9.89949219613813, z=0.0435882),
 ResultRow(mangaid='1-109234', plateifu='9193-1901', ra=45.9974263376, dec=0.426382098814, sersic_logmass=9.71193033196987, z=0.0409309),
 ResultRow(mangaid='1-109250', plateifu='8080-1902', ra=47.7852584272, dec=-1.05434100442, sersic_logmass=8.86787095265526, z=0.0383731),
 ResultRow(mangaid='1-109270', plateifu='9193-12703', ra=46.6858583566, dec=-0.40880014755, sersic_logmass=9.13073892096969, z=0.0252266),
 ResultRow(mangaid='1-109356', plateifu='8081-3703', ra=50.0738407878, dec=0.523632230733, sersic_logmass=9.48148132628646, z=0.0241782),
 ResultRow(mangaid='1-113219', plateifu='7815-9102', ra=317.374745914, dec=10.0519434342, sersic_logmass=9.37199275559893, z=0.0408897),
 ResultRow(mangaid='1-113322', plateifu='7972-12701', ra=315.533514754, dec=11.0751131018, sersic_logmas

## Radial Queries in Marvin
Cone searches can be performed with Marvin Queries using a special **functional** syntax in your SQL string. Cone searches can be performed using the special ``radial`` string function.  The syntax for a cone search query is **radial(RA, Dec, radius)**.  Let's search for all galaxies within 0.5 degrees centered on RA, Dec = 232.5447, 48.6902.  The RA and Dec must be in decimal degrees and the radius is in units of degrees. 

In [15]:
# build the radial filter condition
my_filter = 'radial(232.5447, 48.6902, 0.5)'
q = Query(search_filter=my_filter)
r = q.run()
print(r)
print(r.results)

Marvin Results(query=radial(232.5447, 48.6902, 0.5), totalcount=2, count=2, mode=remote)
<ResultSet(set=1.0/1, index=0:2, count_in_set=2, total=2)>
[ResultRow(mangaid='1-209232', plateifu='8485-1901', ra=232.544703894, dec=48.6902009334),
 ResultRow(mangaid='1-209266', plateifu='8485-9101', ra=233.107502765, dec=48.8332849239)]


## Queries using DAPall parameters.  
MaNGA provides derived analysis properties in its **dapall** summary file.  Marvin allows for queries on any of the parameters in the file.  The table name for these parameters is **dapall**.  Let's find all galaxies that have a total measure star-formation rate > 5 M$_\odot$/year.  The total SFR parameter in the DAPall table is ``sfr_tot``.      

In [17]:
my_filter = 'dapall.sfr_tot > 5'
q = Query(search_filter=my_filter)
r = q.run()
print(r)
print(r.results)

Marvin Results(query=dapall.sfr_tot > 5, totalcount=6, count=6, mode=remote)
<ResultSet(set=1.0/1, index=0:6, count_in_set=6, total=6)>
[ResultRow(mangaid='1-24092', plateifu='7991-1901', sfr_tot=5.5633, bintype_name='VOR10', template_name='GAU-MILESHC'),
 ResultRow(mangaid='1-24092', plateifu='7991-1901', sfr_tot=5.69988, bintype_name='HYB10', template_name='GAU-MILESHC'),
 ResultRow(mangaid='1-37863', plateifu='9193-12704', sfr_tot=6.457, bintype_name='VOR10', template_name='GAU-MILESHC'),
 ResultRow(mangaid='1-37863', plateifu='9193-12704', sfr_tot=6.82693, bintype_name='HYB10', template_name='GAU-MILESHC'),
 ResultRow(mangaid='1-43214', plateifu='8135-1902', sfr_tot=26.4648, bintype_name='VOR10', template_name='GAU-MILESHC'),
 ResultRow(mangaid='1-43214', plateifu='8135-1902', sfr_tot=27.4152, bintype_name='HYB10', template_name='GAU-MILESHC')]


The query returns 6 results, but looking at the plateifu, we see there are only 3 unique targets.  This is because the DAPall file provides measurements for multiple bintypes and by default will return entries for all bintypes.  We can select those out using the ``bintype.name`` parameter. Let's filter on only the HYB10 bintype. 

In [18]:
my_filter = 'dapall.sfr_tot > 5 and bintype.name==HYB10'
q = Query(search_filter=my_filter)
r = q.run()
print(r)
print(r.results)

Marvin Results(query=dapall.sfr_tot > 5 and bintype.name==HYB10, totalcount=3, count=3, mode=remote)
<ResultSet(set=1.0/1, index=0:3, count_in_set=3, total=3)>
[ResultRow(mangaid='1-24092', plateifu='7991-1901', sfr_tot=5.69988, bintype_name='HYB10', template_name='GAU-MILESHC'),
 ResultRow(mangaid='1-37863', plateifu='9193-12704', sfr_tot=6.82693, bintype_name='HYB10', template_name='GAU-MILESHC'),
 ResultRow(mangaid='1-43214', plateifu='8135-1902', sfr_tot=27.4152, bintype_name='HYB10', template_name='GAU-MILESHC')]


## Query on Quality and Target Flags 
Marvin includes the ability to perform queries using quality or target flag information. These work using the special **quality** and **targets** keyword arguments.  These keywords accept a list of flag maskbit labels provided by the [Maskbit Datamodel](https://sdss-marvin.readthedocs.io/en/latest/datamodel/dr15.html#maskbits).  These keywords are inclusive, meaning they will only filter on objects satisfying those labels. 

### Searching by Target Flags
Let's find all galaxies that are in the MaNGA MAIN target selection sample. Targets in the MAIN sample are a part of the PRIMARY, SECONDARY and COLOR-ENHANCED samples.  These are the **primary**, **secondary**, and **color-enhanced** flag labels. The **targets** keywords accepts all labels from the MANGA_TARGET1, MANGA_TARGET2, or MANGA_TARGET3 maskbit schema. 

In [19]:
# create the targets list of labels
targets = ['primary', 'secondary', 'color-enhanced']
q = Query(targets=targets)
r = q.run()
print(r)
print('There are {0} galaxies in the main sample'.format(r.totalcount))
print(r.results[0:5])

Marvin Results(query=None, totalcount=2836, count=100, mode=remote)
There are 2836 galaxies in the main sample
<ResultSet(set=1.0/568, index=0:5, count_in_set=5, total=2836)>
[ResultRow(mangaid='1-109056', plateifu='8077-6103', manga_target1=2080),
 ResultRow(mangaid='1-109073', plateifu='8078-12704', manga_target1=2336),
 ResultRow(mangaid='1-109081', plateifu='8077-12705', manga_target1=4112),
 ResultRow(mangaid='1-109112', plateifu='8078-1901', manga_target1=1040),
 ResultRow(mangaid='1-109152', plateifu='8154-12702', manga_target1=4096)]


The **targets** keyword is equivalent to the ``cube.manga_targetX`` search parameter, where `X` is 1, 2, or 3.  The bits for the primary, secondary, and color-enhanced samples are 10, 11, and 12, respectively.  These combine into the value 7168.  The above query is equivalent to the filter condition ``cube.manga_target1 & 7168`` 

In [20]:
value = 1<<10 | 1<<11 | 1<<12
my_filter = 'cube.manga_target1 & {0}'.format(value)
q = Query(search_filter=my_filter)
r = q.run()
print(r)

Marvin Results(query=cube.manga_target1 & 7168, totalcount=4498, count=100, mode=remote)


Let's search only for galaxies that are ``Milky Way Analogs`` or ``Dwarfs`` ancillary targets. 

In [21]:
targets = ['mwa', 'dwarf']
q = Query(targets=targets)
r = q.run()
print(r)
print('There are {0} galaxies from the Milky Way Analogs and Dwarfs ancillary target catalogs'.format(r.totalcount))
print(r.results)

Marvin Results(query=None, totalcount=26, count=26, mode=remote)
There are 26 galaxies from the Milky Way Analogs and Dwarfs ancillary target catalogs
<ResultSet(set=1.0/1, index=0:26, count_in_set=26, total=26)>
[ResultRow(mangaid='1-121994', plateifu='9485-9102', manga_target3=16384),
 ResultRow(mangaid='1-124604', plateifu='8439-6103', manga_target3=8192),
 ResultRow(mangaid='1-135491', plateifu='9869-6101', manga_target3=8192),
 ResultRow(mangaid='1-146067', plateifu='8937-6102', manga_target3=16384),
 ResultRow(mangaid='1-187821', plateifu='8996-6104', manga_target3=16384),
 ResultRow(mangaid='1-198869', plateifu='8551-6101', manga_target3=16384),
 ResultRow(mangaid='1-199488', plateifu='9038-6103', manga_target3=16384),
 ResultRow(mangaid='1-218519', plateifu='8942-12703', manga_target3=16384),
 ResultRow(mangaid='1-234408', plateifu='8319-6102', manga_target3=8192),
 ResultRow(mangaid='1-23818', plateifu='8611-6104', manga_target3=16384),
 ResultRow(mangaid='1-276557', plateifu=

### Searching by Quality Flags
The **quality** accepts all labels from the MANGA_DRPQUAL and MANGA_DAPQUAL maskbit schema.  Let's find all galaxies that suffered from bad flux calibration.  This is the flag **BADFLUX** (bit 8) from the MANGA_DRPQUAL maskbit schema.  

In [22]:
quality = ['BADFLUX']
q = Query(quality=quality)
r = q.run()
print(r)
print('There are {0} galaxies with bad flux calibration'.format(r.totalcount))
print(r.results[0:10])

Marvin Results(query=None, totalcount=82, count=82, mode=remote)
There are 82 galaxies with bad flux calibration
<ResultSet(set=1.0/9, index=0:10, count_in_set=10, total=82)>
[ResultRow(mangaid='1-109493', plateifu='8156-9101', quality=1073742144),
 ResultRow(mangaid='1-113273', plateifu='7972-1902', quality=1073742144),
 ResultRow(mangaid='1-120967', plateifu='8144-12704', quality=1073742144),
 ResultRow(mangaid='1-121035', plateifu='8144-1901', quality=1073742144),
 ResultRow(mangaid='1-136304', plateifu='8606-1902', quality=1073742144),
 ResultRow(mangaid='1-138157', plateifu='8252-9102', quality=1073742144),
 ResultRow(mangaid='1-149170', plateifu='8997-3701', quality=1073742144),
 ResultRow(mangaid='1-152527', plateifu='8144-6101', quality=1073742144),
 ResultRow(mangaid='1-152769', plateifu='8936-3704', quality=1073742144),
 ResultRow(mangaid='1-174629', plateifu='8947-3703', quality=1073742080)]


The **quality** keyword is equivalent to the search parameters ``cube.quality`` for DRP flags or the ``file.quality`` for DAP flags.  The above query is equivalent to ``cube.quality & 256``.  You can also perform a NOT bitmask selection using the ``~`` symbol.  To perform a NOT selection we can only use the ``cube.quality`` parameter. Let's select all galaxies that do not have bad flux calibration.  

In [23]:
# the above query as a filter condition
q = Query(search_filter='cube.quality & 256')
r = q.run()
print('Objects with bad flux calibration:', r.totalcount)

# objects with bad quality other than bad flux calibration
q = Query(search_filter='cube.quality & ~256')
r = q.run()
print('Bad objects with no bad flux calibration:', r.totalcount)

Objects with bad flux calibration: 82
Bad objects with no bad flux calibration: 604


To find exactly objects with good quality and no bad flags set, use ``cube.quality == 0``. 

In [24]:
q = Query(search_filter='cube.quality == 0')
r = q.run()
print(r)
print('Objects with good quality:', r.totalcount)

Marvin Results(query=cube.quality == 0, totalcount=4253, count=100, mode=remote)
Objects with good quality: 4253


## Useful Resources

Check out these pages on the Marvin Docs site for more information querying with Marvin.

- [Query](https://sdss-marvin.readthedocs.io/en/latest/query/query.html)
- [Query Datamodel](https://sdss-marvin.readthedocs.io/en/latest/datamodel/query_dm.html#query-dm)
- [Results](https://sdss-marvin.readthedocs.io/en/latest/query/results.html)
- [SQL Boolean Syntax Tutorial](https://sdss-marvin.readthedocs.io/en/latest/tutorials/boolean-search-tutorial.html)